### PR TITLE
drivers/sensor: lis2dh: Extended functionality

### PR DESCRIPTION
**Pull request by RubenBoth**
Added support for Wake-up Threshold, and power down functionality as well as some sample functions.

### CODE REVIEW

File: drivers/sensor/lis2dh/Kconfig [hunk 1]
```diff
@@ -138,4 +138,41 @@ config LIS2DH_ODR_9_LOW
...
+	  sample text
+
+config LIS2DH_HPM_OFF
+	bool "No filter"
+
+config LIS2DH_HPM_NORMAL
+	bool "Normal mode"
+
+config LIS2DH_HPM_REFERENCE
+	bool "Difference from reference register"
```
Comment (abelNoelleC) [inline drivers/sensor/lis2dh/Kconfig]: what does this mean?

File: drivers/sensor/lis2dh/lis2dh.c [hunk 2]
```diff
@@ -147,10 +147,49 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
 	odr = lis2dh_freq_to_odr_val(freq);
 	if (odr < 0) {
 		return odr;
+	}else if (odr == 0 && data->powered_down == false){
+		/* check if high resolution mode is enabled as it is not allowed to
```
Comment (abelNoelleC) [inline drivers/sensor/lis2dh/lis2dh.c]: is this the best place to set the low power mode? (if required for the steps in this function, please explain why, otherwise please see if a pm-function can be used)

File: drivers/sensor/lis2dh/lis2dh.c [hunk 3]
```diff
@@ -261,13 +300,105 @@ static const struct sensor_driver_api lis2dh_driver_api = {
...
+		LOG_ERR("Failed to initialize interrupts.");
+		return status;
+	}
+#endif
+
+	LOG_INF("bus=%s fs=%d, odr=0x%x lp_en=0x%x scale=%d",
+		    cfg->bus_name, 1 << (LIS2DH_FS_IDX + 1),
+		    LIS2DH_ODR_IDX, (uint8_t)LIS2DH_LP_EN_BIT, lis2dh->scale);
+
+	/* enable accel measurements and set power mode and data rate */
```
Comment (abelNoelleC) [inline drivers/sensor/lis2dh/lis2dh.c:739]: So are by default all axis used?

File: drivers/sensor/lis2dh/lis2dh.c [hunk 4]
```diff
@@ -261,13 +300,105 @@ static const struct sensor_driver_api lis2dh_driver_api = {
 	.channel_get = lis2dh_channel_get,
 };
 
+int lis2dh_reg_init(const struct device *dev)
```
Comment (abelNoelleC) [inline drivers/sensor/lis2dh/lis2dh.c:669]: why did you create a separate function for this? (so it can be called from the application for something? please specify specific reason in a comment)

File: samples/sensor/lis3dh_test/README.rst [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+********
+
+This sample application periodically reads accelerometer data from the
+LIS2DH sensor (or the compatible LS2DH12, LIS3DH, and LSM303DLHC
+sensors), and displays the sensor data on the console.
+
+Requirements
+************
+
+This sample uses the LIS2DH, ST MEMS system-in-package featuring a 3D
```
Comment (abelNoelleC) [inline samples/sensor/lis3dh_test/README.rst]: specify more specific requirements (only lis2dh, or also other version like lis3dh; certain microcontroller that has specific peripheral pins? etc)

File: samples/sensor/lis3dh_test/README.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+:ref:`actinius_icarus` includes an ST LIS2DH12 accelerometer which
+supports the LIS2DH interface.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/sensor/lis2dh
+   :board: actinius_icarus
+   :goals: build flash
+   :compact:
+
+Building on nucleo_l476rg with IKS01A2 shield
```
Comment (abelNoelleC) [inline samples/sensor/lis3dh_test/README.rst]: do you use this board?

File: drivers/sensor/lis2dh/lis2dh.h [hunk 7]
```diff
@@ -272,5 +321,4 @@ int lis2dh_acc_slope_config(const struct device *dev,
 int lis2dh_spi_init(const struct device *dev);
 int lis2dh_i2c_init(const struct device *dev);
 
-
-#endif /* __SENSOR_LIS2DH__ */
+#endif /* __SENSOR_LIS2DH__ */
```
Comment (abelNoelleC) [inline drivers/sensor/lis2dh/lis2dh.h:327]: check if the blank lines are added to the files where required

File: samples/sensor/lis3dh_test/prj.conf [hunk 8]
```diff
@@ -0,0 +1,13 @@
+CONFIG_STDOUT_CONSOLE=y
```
Comment (abelNoelleC) [inline samples/sensor/lis3dh_test/prj.conf]: check the other merged PRs for an example for specifying which configs are already included in the board config but just added here to be sure, and which ones aren't

**Review by abelNoelleC — COMMENTED** [review]
Added some comments for an initial review, but I'm aware this is WIP (please also add the WIP label in the PR title)

File: drivers/sensor/lis2dh/lis2dh.h (hunk 7 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.h:327]: How would i check this? I put the blank line back for now.

File: samples/sensor/lis3dh_test/README.rst (hunk 5 above)
Comment (RubenBoth) [reply samples/sensor/lis3dh_test/README.rst]: I haven't touched this file yet, it's currently an exact copy of the README from the existing LIS2DH sample application.

File: drivers/sensor/lis2dh/lis2dh.c (hunk 4 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c:669]: I isolated it so it could be re-called after a reset as the reset function only resets the sensors registers and not the device binding etc that is initialized at startup. Do you mean comments in code or in the PR?

File: drivers/sensor/lis2dh/lis2dh.c (hunk 3 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c:739]: Yes, the driver currently doesn't have functionality to select specific axis, so its all or nothing.

File: drivers/sensor/lis2dh/lis2dh.c (hunk 2 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c]: The primary way to select the operation mode (low-power / normal / high res) is done through the kconfig. This PR's added functionality is a way for the application to make the sensor go into a full power down mode. 
When the driver sets the ODR to 0 (which the datasheet describes as 'Power down mode') it tries to save as much power as possible by also putting the sensor in low power mode (regardless of what is selected in the kconfig) and disabling all axis as well. 
When the ODR is then set to a value larger than 0 the driver restores the previous (kconfig selected) operating mode and re-enables the axis. 

This functionality could be put in a dedicated function but this integration felt more streamlined to me. Besides that the datasheet's description of 'power down mode' is tied with the ODR, so retaining that connection in the code felt fitting.

File: drivers/sensor/lis2dh/Kconfig (hunk 1 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/Kconfig]: This is one of the high pass filter modes. I didn't get around to filling it out more as the high pass filter dropped in priority. If i don't get around to it i will remove it.

File: samples/sensor/lis3dh_test/prj.conf (hunk 8 above)
Comment (Duma0) [reply samples/sensor/lis3dh_test/prj.conf:1]: Done

Comment (RubenBoth) [reply samples/sensor/lis3dh_test/prj.conf]: [Commit](https://github.com/abelsensors/zephyr/commit/a583d21a309f806e0aa53a75107c493e0a623df2)
Is this correct?

File: drivers/sensor/lis2dh/lis2dh.c [hunk 9]
```diff
@@ -124,9 +124,9 @@ static int lis2dh_freq_to_odr_val(uint16_t freq)
 	size_t i;
 
 	/* An ODR of 0 Hz is not allowed */
-	if (freq == 0U) {
-		return -EINVAL;
-	}
+	//if (freq == 0U) {
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c]: Please remove comment

File: drivers/sensor/lis2dh/Kconfig (hunk 1 above)
Comment (abelNoelleC) [reply drivers/sensor/lis2dh/Kconfig]: For the time being I would advise to either remove it or to clearly indicate that this is WIP with a comment or something like that

File: drivers/sensor/lis2dh/lis2dh.c (hunk 2 above)
Comment (abelNoelleC) [reply drivers/sensor/lis2dh/lis2dh.c]: As we discussed in the meeting on Tuesday, I'm still not sure if doing it like this is the best way. The drivers have device power management capabilities, and if you still have time it would be very beneficial to look into this. Also including stuff like turning off axes, which can greatly impact the power consumption (but this could potentially fit better in the init function).
I understand that the ODR and power down mode are interlinked, so if possible please think about the best solution keeping also the above things in mind

File: drivers/sensor/lis2dh/lis2dh.c (hunk 3 above)
Comment (abelNoelleC) [reply drivers/sensor/lis2dh/lis2dh.c:739]: Please state this clearly in a comment and add that as a todo for the driver (somewhere where you're creating an overview of the functionalities and limitations of the driver's current state, like at the beginning of the driver or the README etc)

File: drivers/sensor/lis2dh/lis2dh.c (hunk 4 above)
Comment (abelNoelleC) [reply drivers/sensor/lis2dh/lis2dh.c:669]: Comment in code

File: samples/sensor/lis3dh_test/README.rst (hunk 5 above)
Comment (abelNoelleC) [reply samples/sensor/lis3dh_test/README.rst]: Thought so :), then see this as a reminder to not forget it ;)

File: drivers/sensor/lis2dh/lis2dh.h (hunk 7 above)
Comment (abelNoelleC) [reply drivers/sensor/lis2dh/lis2dh.h:327]: if you go to the 'Files Changed' tab in the PR, you'll see some red symbols (circle with a stripe) at the end of some files, indicating the lack of a blank line

File: samples/sensor/lis3dh_test/prj.conf (hunk 8 above)
Comment (abelNoelleC) [reply samples/sensor/lis3dh_test/prj.conf]: Yes, maybe also make a section for debugging/logging stuff if you're going to separate that from the other two, but this is what I meant

**Review by abelNoelleC — COMMENTED** [review]
Responded to some comments

File: drivers/sensor/lis2dh/lis2dh.c (hunk 2 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c]: The power down functionality is now moved to a dedicated function. The option to enable/disable individual axis is on my todo list.

File: drivers/sensor/lis2dh/lis2dh.c [hunk 10]
```diff
@@ -83,7 +83,7 @@ static int lis2dh_sample_fetch(const struct device *dev,
 {
 	struct lis2dh_data *lis2dh = dev->data;
 	size_t i;
-	int status;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:86]: Is there a reason you changes status to ret?  Does it serve a purpose?

File: drivers/sensor/lis2dh/lis2dh.c [hunk 11]
```diff
@@ -114,7 +113,7 @@ static int lis2dh_sample_fetch(const struct device *dev,
 	return -ENODATA;
 }
 
-#ifdef CONFIG_LIS2DH_ODR_RUNTIME
+#if defined(CONFIG_LIS2DH_ODR_RUNTIME)
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:116]: Again why did you change it ? The reason I am asking is the less changes you have the less comments you will get from the community

File: drivers/sensor/lis2dh/lis2dh.c [hunk 12]
```diff
@@ -133,25 +134,25 @@ static int lis2dh_freq_to_odr_val(uint16_t freq)
...
 }
 
 static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
 {
 	int odr;
-	int status;
+	int ret;
 	uint8_t value;
-	struct lis2dh_data *data = dev->data;
+	struct lis2dh_data *lis2dh = dev->data;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:145]: Try to do only relevant changes

File: drivers/sensor/lis2dh/lis2dh.c [hunk 13]
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
...
 }
 #endif
 
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
+#if (defined(CONFIG_LIS2DH_OPER_MODE_RUNTIME) || defined(CONFIG_LIS2DH_OPER_MODE_HIGH_RES))
+static int lis2dh_set_high_res(const struct device *dev, bool enable)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	int ret;
+	uint8_t value;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:196]: Always initialize your variables

File: drivers/sensor/lis2dh/lis2dh.c [hunk 14]
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
...
+
+	if (new_mode == OPER_MODE_LOW_POWER){
+		if (lis2dh->oper_mode == OPER_MODE_HIGH_RES) {
+			ret = lis2dh_set_high_res(dev, false);
+			if (unlikely(ret < 0)) {
+				return ret;
+			}
+		}
+		ret = lis2dh->hw_tf->update_reg(dev, LIS2DH_REG_CTRL1,
+									(uint8_t)LIS2DH_LP_EN_BIT_MASK,
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c]: I would take care on your formatting a bit. This extra tabs are a bit too much

File: drivers/sensor/lis2dh/lis2dh.c [hunk 15]
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
...
+#endif
+
+#if defined(CONFIG_PM_DEVICE)
+static int lis2dh_suspend(const struct device *dev)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	if (lis2dh->pm_state == PM_DEVICE_STATE_ACTIVE) {
+#if defined(CONFIG_LIS2DH_OPER_MODE_HIGH_RES)
+		int ret;
+		uint8_t value;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c]: Again some initialization

File: drivers/sensor/lis2dh/lis2dh.c [hunk 16]
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
...
+	} 
+}
+
+static int lis2dh_resume_from_suspend(const struct device *dev)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+
+	if (lis2dh->pm_state == PM_DEVICE_STATE_LOW_POWER) {
+		
+		uint8_t axes = 0;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:311]: Good one :)

File: drivers/sensor/lis2dh/lis2dh.c [hunk 17]
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
...
 
+int lis2dh_acc_act_config(const struct device *dev,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	int ret;
+
+	if ((uint16_t)attr == SENSOR_ATTR_LIS2DH_ACT_TH) {
+		uint8_t range_g, reg_val;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c]: Always try to initialize

File: drivers/sensor/lis2dh/lis2dh.c [hunk 18]
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
...
+						 &reg_val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* fs reg value is in the range 0 (2g) - 3 (16g) */
+		range_g = 2 * (1 << ((LIS2DH_FS_MASK & reg_val)
+				      >> LIS2DH_FS_SHIFT));
+
+		act_th_ums2 = val->val1 * 1000000 + val->val2;
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:424]: Add a comment why you are doing this

File: drivers/sensor/lis2dh/lis2dh.c [hunk 19]
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
...
+}
+
 static int lis2dh_acc_config(const struct device *dev,
 			     enum sensor_channel chan,
 			     enum sensor_attribute attr,
 			     const struct sensor_value *val)
 {
-	switch (attr) {
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
+	switch ((uint16_t)attr) {
```
Comment (abelkarapatis) [inline drivers/sensor/lis2dh/lis2dh.c:460]: Is there a reason you are casting?  The enum is by default an int value

**Review by abelkarapatis — COMMENTED** [review]
Nice work

File: drivers/sensor/lis2dh/lis2dh.c (hunk 10 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c:86]: Some functions use ret, others used status. I changed all status to ret to make the code more consistent. It doesn't serve any functional purpose.

File: drivers/sensor/lis2dh/lis2dh.c (hunk 11 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c:116]: Mostly code consistency.

File: drivers/sensor/lis2dh/lis2dh.c (hunk 14 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c]: My IDE displays tabs shorter then github. Ill remove some tabs.

File: drivers/sensor/lis2dh/lis2dh.c (hunk 19 above)
Comment (RubenBoth) [reply drivers/sensor/lis2dh/lis2dh.c:460]: This has to do with adding custom sensor attributes. If the value is not casted it cannot be compared to two different enumerators

Comment (abelkarapatis) [reply drivers/sensor/lis2dh/lis2dh.c:460]: Then my advice would be not to use 2 different enumerators as this can lead to a lot of confusion. Try to place every in one enum. Having an enum helps the compiler check which cases there should be

File: samples/sensor/lis3dh_test/boards/stm32l1_disco.overlay [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+    pinctrl-0 = <&i2c1_scl_pb6 &i2c1_sda_pb7>;
+
+    lis3dh: lis2dh@18 {
+        status = "okay";
+        compatible = "st,lis2dh";
+        label = "LIS3DH";
+        reg = <0x18>;
+    	irq-gpios= <&gpiob 8 GPIO_ACTIVE_HIGH>, <&gpiob 8 GPIO_ACTIVE_HIGH>;
+    };
+};
```
Comment (abelNoelleC) [inline samples/sensor/lis3dh_test/boards/stm32l1_disco.overlay:13]: mind the missing blank lines at the end of these files (check the red circle with stripe)

File: samples/sensor/lis3dh_test/prj.conf [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+CONFIG_LIS2DH_ODR_RUNTIME=y
+CONFIG_LIS2DH_ACCEL_RANGE_16G=y
+CONFIG_LIS2DH_OPER_MODE_RUNTIME=y
+CONFIG_LIS2DH_AXES_RUNTIME=y
+CONFIG_PM_DEVICE=y
+
+### Debugging ###
+CONFIG_LOG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
+CONFIG_CBPRINTF_FP_SUPPORT=y
```
Comment (abelNoelleC) [inline samples/sensor/lis3dh_test/prj.conf:18]: missing blank line

**Review by abelNoelleC — COMMENTED** [review]
Please let us know when the PR is ready for a more thorough review, for now it's just two small comments that I forgot to actually submit :)
